core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -237,601 +220,1206 @@ private SqlQueryScheduler(
+            this.taskLifecycleListener = requireNonNull(taskLifecycleListener, "taskLifecycleListener is null");
+            this.stages = requireNonNull(stages, "stages is null");
+            this.executions = ImmutableList.copyOf(requireNonNull(executions, "executions is null"));
+            this.dynamicFilterService = requireNonNull(dynamicFilterService, "dynamicFilterService is null");
+            this.distributedStagesScheduler = requireNonNull(distributedStagesScheduler, "distributedStagesScheduler is null");
+            this.coordinatorTaskManager = requireNonNull(coordinatorTaskManager, "coordinatorTaskManager is null");
+        }
+
+        private void initialize()
+        {
+            for (PipelinedStageExecution execution : executions) {
+                execution.addStateChangeListener(state -> {
+                    if (queryStateMachine.isDone()) {
+                        return;
+                    }
+                    if (!state.canScheduleMoreTasks()) {
+                        dynamicFilterService.stageCannotScheduleMoreTasks(execution.getStageId(), execution.getAllTasks().size());
+                    }
+                    // if any coordinator stage failed transition directly to failure
+                    if (state == FAILED) {
+                        RuntimeException failureCause = execution.getFailureCause()
+                                .map(ExecutionFailureInfo::toException)
+                                .orElseGet(() -> new VerifyException(format("stage execution for stage %s is failed by failure cause is not present", execution.getStageId())));
+                        stages.get(execution.getStageId()).fail(failureCause);
+                        queryStateMachine.transitionToFailed(failureCause);
+                    }
+                    else if (state == ABORTED) {
+                        // this should never happen, since abort can only be triggered in query clean up after the query is finished
+                        stages.get(execution.getStageId()).abort();
+                        queryStateMachine.transitionToFailed(new TrinoException(GENERIC_INTERNAL_ERROR, "Query stage was aborted"));
+                    }
+                    else if (state.isDone()) {
+                        stages.get(execution.getStageId()).finish();
+                    }
+                });
+            }
+
+            for (int currentIndex = 0, nextIndex = 1; nextIndex < executions.size(); currentIndex++, nextIndex++) {
+                PipelinedStageExecution current = executions.get(currentIndex);
+                PipelinedStageExecution next = executions.get(nextIndex);
+                verify(getOnlyElement(stages.getChildren(current.getStageId())).getStageId().equals(next.getStageId()), "coordinator stages are expected to consume only a singe other coordinator stage");
```

Comment:
nit: typo singe

Reply:
Also the error message does not capture condition. It caputres the expectation for children to be single element list, and that is verified via call to `getOnlyElement`.
Her you should say that sages ordering does not match linking between stages.